### PR TITLE
explore_text_columns

### LINKED ISSUE #1: explore_text_columns - Write Specification
### Function: `explore_text_columns`

- [x] For this milestone you will write R function documentation that will serve as specifications, but you will NOT write any code for your package functions. Specifically, set-up empty functions (with appropriate function names) containing no code in the files where you will eventually write code. Write function documentation (e.g., complete roxygen2 comments). These are the functions you wrote about in your proposal.

### LINKED ISSUE #10: explore_text_columns - Write Function internals and test cases
1. Write test cases and code iteratively¶

rubric={accuracy:20,quality:10,mechanics:10}

   - Before writing any code, revisit your function specifications from last week and revise them if needed. As you do this, think critically about what function inputs you expect from the user and what your function is supposed to do (e.g., return).

   - Write a unit tests for each specified function using testthat. test_that code blocks do not need to be wrapped in a function. Each function should have 3 to 5 edge cases tested to ensure that the function returns what is expected to the user. Group these test cases in a sensible way so that what is being tested matches the string associated with the test_that code blocks. This might mean that you may end up with 2-3 test_that code blocks per function. Remember, when writing your tests, build them around your function specifications and requirements.

-    After writing your unit tests, write the internal code for your R functions.

 Write function code and tests in close temporal proximity with each other. The process of writing function code and test cases will be iterative; there will be several rounds of writing tests →function improvement → writing more tests → function improvement → writing more tests →function improvement, etc.

- Make sure that the tests cover all branches (e.g., if your function has an if statement, you should have tests where the if statement is true, and where the if statement is false.

 - As you develop your code and test cases, update your code documentation so that it makes sense with any of the changes you have made. Do not forget to also document your test functions.

 - Check whether the code passes all the unit tests you have written using the devtools::test() function.

### Exception handling¶

rubric={quality:10}

    The function should be written defensively. That is, it should handle incorrect input and errors detected during execution via throwing exceptions with useful error messages, and there should be tests to confirm that the exceptions result in the expected behaviour of the function.

### PR DESCRIPTION
**Pull request by charlessuresh**
- Wrote test cases and coded iteratively
- added exception handling
- updated function specification
- updated `explore_text_columns` section of README
- tested function in `eda_playground.Rmd`

The functions `explore_text_columns` and `test-explore_text_columns` are a bit long. However, because of time limitations, I don't think I can break them up into smaller functions in time for Milestone 3. I will probably reserve that for Milestone 4

@lephanthuymai please review
@rahulkuriyedath @adibns feel free to review as well

### CODE REVIEW

File: tests/testthat/test-explore_text_columns.R [hunk 1]
```diff
@@ -1,3 +1,47 @@
+require(tidyverse)
+require(vdiffr)
+
 test_that("multiplication works", {
```
Comment (lephanthuymai) [inline tests/testthat/test-explore_text_columns.R]: Please change the description of the test.

File: R/explore_text_columns.R [hunk 2]
```diff
@@ -1,19 +1,311 @@
...
+  if (!is.character(text_cols)) {
+    stop("text_cols is not passed as a character vector")
+  }
+
+  if (!is.data.frame(df)) {
+    stop("The df passed is not a Data Frame")
+  }
+
+  if(length(text_cols) == 0) {
+    for (col in colnames(df %>% dplyr::select_if(is.character))){
```
Comment (lephanthuymai) [inline R/explore_text_columns.R:46]: I would suggest adding some comments here on how you retrieve the text columns.

File: R/explore_text_columns.R [hunk 3]
```diff
@@ -1,19 +1,311 @@
...
+      cat("Identified the following as text columns:", text_cols)
+      cat("\n")
+      cat("\n")
+    }
+  } else {
+    for (col in text_cols){
+      if (!(col %in% colnames(df))){
+        stop(paste0(col, " passed in text_cols is not a column in the dataframe"))
+      }
+      if (!is.character(df)){
```
Comment (lephanthuymai) [inline R/explore_text_columns.R]: was this meant to be !is.character(col) instead of !is.character(df)?

File: R/explore_text_columns.R [hunk 4]
```diff
@@ -1,19 +1,311 @@
...
-#' @return a list of plots
+#' @return a list of results and plots
+#' @export
 #'
 #' @examples
-#' \dontrun{
-#' explore_text_columns(df)
-#' }
-explore_text_columns <- function(df, text_cols=NULL){
+#' explore_text_columns(cars)
```
Comment (lephanthuymai) [inline R/explore_text_columns.R:25]: should we give another example? as there is no text column in cars

**Review by lephanthuymai — COMMENTED** [review]
Overall, great work, Charles!
I feel like there is a lack of comments in your code, adding some describing what will happen in each part of your function will be helpful for others to review your code (especially the TA).
Thank you for your hard work on this.
Best,
Mai

File: tests/testthat/test-explore_text_columns.R (hunk 1 above)
Comment (charlessuresh) [reply tests/testthat/test-explore_text_columns.R]: Thanks for pointing this out!
I've addressed this in this [commit](https://github.com/UBC-MDS/datascience.eda.R/pull/15/commits/b0ae900e78f9028610288ce5fe56634ae825dac8)

File: R/explore_text_columns.R (hunk 2 above)
Comment (charlessuresh) [reply R/explore_text_columns.R:46]: I've added comments to the entire code now
I've addressed this is this [commit](https://github.com/UBC-MDS/datascience.eda.R/pull/15/commits/84f285d351b363b8399d26529026ad0be8e11c7a)

File: R/explore_text_columns.R (hunk 3 above)
Comment (charlessuresh) [reply R/explore_text_columns.R]: Good spot!
I've addressed this is in this [commit](https://github.com/UBC-MDS/datascience.eda.R/pull/15/commits/3ca7a07b6426286a79a5b8b44515395ebef1f3ff)

File: R/explore_text_columns.R (hunk 4 above)
Comment (charlessuresh) [reply R/explore_text_columns.R:25]: I've added another example using the "apocrypha" dataset from a developer package [JohnCoene/sacred](https://github.com/JohnCoene/sacred)
I've addressed this is in this [commit](https://github.com/UBC-MDS/datascience.eda.R/pull/15/commits/5d271e4a9a0db4637edfa95e1f331eb4d6a8285a)

**Comment by charlessuresh** [comment]
Hey @lephanthuymai 

Thanks for the detailed and in-depth review. I've addressed the issues pointed out via the latest commits.
Request you to review the corrections/additions made.

**Comment by adibns** [comment]
Nothing much to say from my end. Awesome work @charlessuresh
